xds/src/main/java/io/grpc/xds/sds/TlsCertificateStore.java

```diff
@@ -0,0 +1,89 @@
+/*
+ * Copyright 2019 The gRPC Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc.xds.sds;
+
+import com.google.common.base.Preconditions;
+import com.google.protobuf.ByteString;
+import io.envoyproxy.envoy.api.v2.auth.TlsCertificate;
+import io.envoyproxy.envoy.api.v2.core.DataSource;
+import io.grpc.Internal;
+import java.io.InputStream;
+
+/**
+ * TlsCertificate's PrivateKey and Certificate are extracted into InputStream's.
+ * This is used by the gRPC SSLContext/Protocol Negotiator and is internal
+ */
+@Internal
+public final class TlsCertificateStore {
+
+  private final InputStream privateKeyStream;
+  private final InputStream certChainStream;
+
+  private InputStream getInputStreamFromDataSource(DataSource dataSource) {
+    Preconditions.checkNotNull(dataSource);
+    ByteString dataSourceByteString = null;
+    if (dataSource.getSpecifierCase() == DataSource.SpecifierCase.INLINE_BYTES) {
+      dataSourceByteString = dataSource.getInlineBytes();
+    } else if (dataSource.getSpecifierCase() == DataSource.SpecifierCase.INLINE_STRING) {
+      dataSourceByteString = dataSource.getInlineStringBytes();
+    } else {
+      throw new IllegalArgumentException(
+          "dataSource of type " + dataSource.getSpecifierCase() + " not supported");
+    }
+    return dataSourceByteString.newInput();
+  }
+
+  /**
+   * Creates the Store out of the TlsCertificate object of xDS.
+   *
+   * @param tlsCertificate  TlsCertificate Object of xDS
+   */
+  public TlsCertificateStore(TlsCertificate tlsCertificate) {
+    Preconditions.checkNotNull(tlsCertificate);
+    privateKeyStream = getInputStreamFromDataSource(tlsCertificate.getPrivateKey());
+    certChainStream = getInputStreamFromDataSource(tlsCertificate.getCertificateChain());
+  }
+
+  /**
+   * Creates the Store out of 2 streams for the 2 certs on disk.
+   *
+   * @param privateKeySteam  stream representing private key on disk
+   * @param certChainStream  stream representing cert on disk
+   */
+  public TlsCertificateStore(InputStream privateKeySteam, InputStream certChainStream) {
+    this.privateKeyStream = privateKeySteam;
+    this.certChainStream = certChainStream;
+  }
+
+  /**
+   * getter for private key stream.
+   *
+   * @return inputStream representing private key
+   */
+  public InputStream getPrivateKeyStream() {
```

Comment:
When is the input stream supposed to be closed? Who manages its life cycle?

Reply:
Lifecyle management including refCounting and removal is coming in another PR

Reply:
Sure. If you know how the lifecycle management will be can you add some javadoc?

Reply:
Done here and main note in SecretManager

Reply:
can you explain briefly about how you will handle in other PR? to me this is one of the most concerning part of this PR which may ends up changing some structure depends on how you are planning to handle the lifecycle / ownership of the stateful resource.

Reply:
Okay, based on offline discussion - the lifecycle management of the object is what I referred to earlier.  The ownership/state management of the stream is the concern here which I will address now.